### PR TITLE
Rework build and add to github actions

### PR DESCRIPTION
**Pull request by JamesGuthrie**

### CODE REVIEW

File: Makefile
Comment (bitwalker) [inline Makefile:28]: **Suggestion:** We probably want to use buildx for all of our image builds, since it has all of the rich features, including multi-arch images, which `docker build` doesn't support (currently anyway). It's just `docker buildx build` instead, which is basically identical with `docker build`, but also supports all of the extra functionality that buildx provides. It's especially important in order for caching work in GitHub Actions.

I maintain a handful of Docker images which use buildx and build/publish via GitHub Actions, so if you'd like some references to look at, just hit me up on Slack.

**Review by bitwalker — COMMENTED** [review]
Overall looks good, but I think we can take this opportunity to simplify the Dockerfile considerably. I've suggested a few changes that I think we should make, though nothing that blocks this PR from being merged as-is if you'd prefer to make those changes in new PRs.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile:38]: I honestly can't answer this. I suspect that @cevian would prefer that we don't remove it.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile:48]: I will remove sccache from this change for now. I experimented with a bunch of different things (including caching the target directory), but I found that the sccache approach was the least "invasive".

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile:60]: Previously we were installing rust as root and then chmodding the rust install etc. It just seemed weirdly fragile to me, and I had some issues when trying to debug stuff in the half-built container (which I can't remember exactly anymore 😅 ). That's why I refactored this to install rust as postgres.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile:69]: @cevian can you comment on why we took this approach in the first place? I would be for simplifying this step.

File: Makefile
Comment (JamesGuthrie) [reply Makefile:28]: https://github.com/timescale/promscale_extension/pull/52

File: Makefile
Comment (JamesGuthrie) [reply Makefile:1]: 👍

File: Dockerfile
Comment (bitwalker) [reply Dockerfile:38]: @cevian Do you know what purpose pgextwlist serves in this Dockerfile?

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile:67]: Unfortunately it seems as though removing this breaks things.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: done

File: Makefile
Comment (JamesGuthrie) [reply Makefile]: done

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: done

File: Dockerfile
Comment (bitwalker) [reply Dockerfile:54]: Could you paste the error you observed or what was broken? That feature is enabled by default on musl targets, so setting this should be a no-op anyway.

Comment (JamesGuthrie) [reply Dockerfile:54]: So the issue that I saw is that after installing `cargo-pgx`, the `cargo pgx init ...` command failed with the following error:
```
could not execute process `/var/lib/postgresql/.cargo/bin/cargo-pgx pgx init --pg14 /usr/local/bin/pg_config` (never executed)
```

more context:

```
#9 141.8   Installing /var/lib/postgresql/.cargo/bin/cargo-pgx
#9 141.8    Installed package `cargo-pgx v0.2.6` (executable `cargo-pgx`)
#9 142.0 + which pg_config
#9 142.0 + cargo pgx init --pg14 /usr/local/bin/pg_config
#9 142.0 error: could not execute process `/var/lib/postgresql/.cargo/bin/cargo-pgx pgx init --pg14 /usr/local/bin/pg_config` (never executed)
#9 142.0
#9 142.0 Caused by:
#9 142.0   No such file or directory (os error 2)```

Comment (bitwalker) [reply Dockerfile:54]: I'm going to pull your branch quick and try to see what the underlying issue is, something is definitely wonky there.

**Comment by bitwalker** [comment]
@JamesGuthrie I pushed a commit that implements some of my suggestions above, feel free to rewrite the commit or `git reset --hard HEAD^` on your branch if you decide you don't want it as-is, but figured it was easier to push it to share the final set of changes I landed on. I left some notes on our other comment threads expanding on some things I found, but I think the result is much simpler now.

File: Dockerfile
Comment (bitwalker) [reply Dockerfile]: After removing it, I'm not seeing any differences as a result, so I'm pretty convinced it isn't serving any purpose currently.

File: Dockerfile
Comment (bitwalker) [reply Dockerfile]: So it turns out that the issue is that `initdb` cannot be run as root, which I forgot about, so that means you need to run `cargo pgx init` as another user (i.e. `postgres`), but if you wish to do this without changing anything else, there are problems:

* Rustup is intended to be run by each user on the system that wants Rust, and many paths are relative to the current user's `HOME`, e.g. `CARGO_HOME`, `RUSTUP_HOME`, and `PGX_HOME`. 
* If you just change `USER` in the Dockerfile, `HOME` doesn't change, and the permissions will all be that of whatever user installed Rust (root in this case).

In the base image we're using, `HOME` is set to `/root`, and `HOME` for the `postgres` user is set to `/var/lib/postgresql` - but `USER postgres` doesn't change the value of `HOME` for subsequent `RUN` commands, it only takes effect when running the final image. As a result, we either need to set `HOME` to something; install things in `/var/lib/postgresql` (which feels gross); or create a directory we can use like `HOME` using the various `*_HOME` env vars that Rust and friends respect. I chose the first one for simplicity.

File: Dockerfile
Comment (bitwalker) [reply Dockerfile]: It turns out I misread the value of `RUSTFLAGS` here, we're actually _disabling_ the `crt-static` feature, not enabling it (since that is the default, and would be represented as `+crt-static` if we were enabling it manually). This is needed in order to build dylibs on Alpine (which obviously we require for the extension). I've added a comment to that effect.

File: Dockerfile
Comment (bitwalker) [reply Dockerfile]: I went ahead and changed this to the simplest version, so unless we have a good reason for making it more complicated, I'd suggest we keep it that way.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: I'm still hitting the same issue as before with your latest change. Perhaps it's an arm64 thing?

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: Promscale's `sql_permissions.md` mentions the following:

# Using a non-superuser owner with pgextwlist

The owner user of Promscale need to be able to install
and upgrade the promscale and timescaledb PostgreSQL extensions.

Normally extension installation and upgrade requires superuser permissions. To avoid
using superuser, you could use the [pgextwlist](https://github.com/dimitri/pgextwlist) extension with the following PostgreSQL
config:
```
local_preload_libraries=pgextwlist
extwlist.extensions=promscale,timescaledb
```

File: Dockerfile
Comment (bitwalker) [reply Dockerfile]: That's strange, must not be related to the `RUSTFLAGS` then because it wasn't removed in my changes. I can't reproduce on x86_64 (I use a desktop machine in my house as my Docker builder, rather than my macbook). I'll try and do a build locally on my laptop and see if I can figure out what's going on.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: 👍

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: But your change does move `RUSTFLAGS` to be set _after_ pgx install.

Comment (bitwalker) [reply Dockerfile]: Ah, you know, I did move it down to above our `make package` invocation, perhaps it is needed by `cargo install pgx` or something. I'll move it up to above all the cargo builds.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: LGTM.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: I know that we were planning on changing this by marking the extension as trusted.

File: Dockerfile
Comment (JamesGuthrie) [reply Dockerfile]: The latest commit fixes this 👍

File: Dockerfile
Comment (bitwalker) [reply Dockerfile]: We're installing it in the build stage though, not the final image that users will use. If we do want to support this though, I'd suggest we add it to the base timescaledb image, since that feels more appropriate rather than having it only in the image that has the Promscale extension.

Comment (JamesGuthrie) [reply Dockerfile]: doh. I didn't even realise that. I may have broken that in my refactoring.

Comment (bitwalker) [reply Dockerfile]: It looks like pgextwlist _would_ be in the final image in the original Dockerfile, because we were copying all of `/usr/local/lib/postgresql` and `/usr/local/share/postgresql` into the final image, and I'm assuming pgextwlist was being installed in those. With the changes made to copy from the directories generated by `pgx` though, which don't contain that extension AFAICT, they are no longer being copied into the final image.

Comment (JamesGuthrie) [reply Dockerfile]: Yup, I just verified that I broke that. pgextwlist.so was present on the image before this refactoring.

Comment (bitwalker) [reply Dockerfile]: Since we're planning to mark the extension as trusted anyway, it seems we probably don't need it, at least based on the docs. @cevian do you know if there is any other reason why we would still require pgextwlist in that scenario?

File: Makefile
Comment (paulfantom) [inline Makefile:1]: Why not default to `all` and allow easy inclusion of other potential targets to simplify development (like running tests, cleaning env before build, etc.) ?

Comment (JamesGuthrie) [reply Makefile:1]: I'm not sure there's an added value of an `all` at the moment. The way that cargo works, there's no added value in trying to build incremental steps on top of it.

I also don't use `make` for development. I do see the value of the Makefile in giving people new to the project a clear entrypoint, but aside from that it doesn't really do much except wrap cargo commands.

Comment (paulfantom) [reply Makefile:1]: ACK, makes sense :+1:

Comment (bitwalker) [reply Makefile:1]: I put something like this in all my Makefiles, occasionally tailored to the specific project, and use it as the default target so when you run `make` or `make help`, you get a list of all the targets with docs:

```makefile
IMAGE_NAME ?= project/image
VERSION ?= `cat VERSION`

.PHONY: help
help:
	@echo "$(IMAGE_NAME):$(VERSION)"
	@perl -nle'print $& if m{^[a-zA-Z_-]+:.*?## .*$$}' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'

.PHONY: example_target
example_target: ## Some description of what this target does
```

**Comment by bitwalker** [comment]
@JamesGuthrie I suggest we go ahead and merge this as-is, and if we need to make further adjustments, they can happen in follow-on PRs, what do you think?

File: Dockerfile
Comment (cevian) [reply Dockerfile]: pgextwlist should be necessary for some tests from the connector and is necessary for the permission instructions cited above. Since we aren't marking the extension trusted yet, I don't think we can remove it. Although after it's trusted then we can

Comment (bitwalker) [reply Dockerfile]: Makes sense, I was under the impression we aren't planning another release of the extension until our other changes are in place though, including marking the extension trusted, so we should be in the clear to remove it. The previous docker images containing pgextwlist and the latest release of the extension are still available (and presumably used by promscale for its test suite), so these changes only affect current development and future releases.

**Review by paulfantom — APPROVED** [review]
CI and build files look good to me.

Small, non-blocking question - do you want to publish somewhere images built in CI?

**Comment by JamesGuthrie** [comment]
> do you want to publish somewhere images built in CI

At some point we will add that, yes.

**Comment by JamesGuthrie** [comment]
@bitwalker you're right. Let's merge this for now. We will figure out the trusted extension stuff along the way anyway. I opened #54 to keep track of that.

**Pull request merged**
